src/Chainweb/Pact/Backend/DbCache.hs

```diff
@@ -193,12 +230,60 @@ cacheStats mc = object
 updateCacheStats :: DbCache a -> (Value, DbCache a)
 updateCacheStats mc = (cacheStats mc, set dcMisses 0 (set dcHits 0 mc))
 
+unionDbCache :: DbCache a -> DbCache a -> DbCache a
+unionDbCache c c' = DbCache
+    { _dcStore = HM.union (_dcStore c) (_dcStore c')
+    , _dcSize = _dcSize c + _dcSize c'
+    , _dcLimit = max (_dcLimit c) (_dcLimit c')
+    , _dcMisses = _dcMisses c + _dcMisses c'
+    , _dcHits = _dcHits c + _dcHits c'
+    }
+
+filterDbCache :: (CacheAddress -> CacheEntry a -> Bool) -> DbCache a -> DbCache a
+filterDbCache f DbCache{..} = DbCache
+    { _dcStore = m
+    , _dcSize = HM.size m
+    , _dcLimit = _dcLimit
+    , _dcMisses = 0
+    , _dcHits = 0
+    }
+    where m = HM.filterWithKey f _dcStore
+
+fromHashMap :: TxId -> HM.HashMap ModuleName (ModuleData Ref,Bool) -> DbCache PersistModuleData
+fromHashMap txid m = (emptyDbCache defaultModuleCacheLimit)
+    { _dcStore = HM.fromList $ map convert $ HM.toList m
+    , _dcSize = HM.size m
+    }
+    where
+        toPersist :: ModuleData Ref -> PersistModuleData
+        toPersist d = either (const $ error "fromHashMap: toPersist failed") id (traverse (traverse toPersistDirect) d)
+
+        convert :: (ModuleName, (ModuleData Ref,Bool)) -> (CacheAddress, CacheEntry PersistModuleData)
+        convert (ModuleName n _, (d,_)) =
+            let ca = CacheAddress $ BS.toShort $ TE.encodeUtf8 n
+            in (ca, CacheEntry txid ca 1 (toPersist d) 0)
+
+toHashMap :: DbCache PersistModuleData -> HM.HashMap ModuleName (ModuleData Ref,Bool)
+toHashMap DbCache{_dcStore} = HM.fromList $ map convert $ HM.toList filteredStore
+    where
+        filteredStore = HM.filterWithKey (\(CacheAddress ca) _ -> not $ isInfixOf "hashed:" (BS.fromShort ca)) _dcStore
+
+        fromPersist :: PersistModuleData -> ModuleData Ref
+        fromPersist d = either (const $ error "toHashMap: fromPersist failed") id (traverse (traverse (fromPersistDirect nativeLookup)) d)
+
+        convert :: (CacheAddress, CacheEntry PersistModuleData) -> (ModuleName, (ModuleData Ref,Bool))
+        convert (CacheAddress ca, CacheEntry{_ceData}) =
+            (ModuleName (TE.decodeUtf8 $ BS.fromShort ca) Nothing, (fromPersist _ceData, False))
+
+keysDbCache :: DbCache PersistModuleData -> [CacheAddress]
+keysDbCache DbCache{_dcStore} = HM.keys _dcStore
+
 -- -------------------------------------------------------------------------- --
 -- Internal
 
 mkAddress :: Utf8 -> ByteString -> CacheAddress
 mkAddress (Utf8 key) rowdata = CacheAddress . BS.toShort $
-    key <> ":" <> BA.convert (C.hash @_ @SHA512t_256 rowdata)
+    "hashed:" <> key <> ":" <> BA.convert (C.hash @_ @SHA512t_256 rowdata)
```

Comment:
why this change?

Reply:
We have to filter out these keys as they consist of non-unicode hashes while `ModuleName` type is a `Text`, and they are very specific.